pkg/webhook/policy.go

```diff
@@ -430,3 +410,12 @@ func (h *validationHandler) tracingLevel(ctx context.Context, req admission.Requ
 func (h *validationHandler) skipExcludedNamespace(namespace string) bool {
 	return h.processExcluder.IsNamespaceExcluded(process.Webhook, namespace)
 }
+
+func getViolationRef(gkNamespace, rkind, rname, rnamespace, ckind, cname, cnamespace string) *corev1.ObjectReference {
+	return &corev1.ObjectReference{
+		Kind:      rkind,
+		Name:      rname,
+		UID:       types.UID(rkind + rname + rnamespace + ckind + cname + cnamespace),
```

Comment:
same comments about GKNN and separators here

Also, should webhooks be using the resource kind/name as a key in the object reference?

This would allow a user who technically doesn't have permissions to create objects flood the API server with events by spamming the webhook.

Reply:
Wouldn't the authorization check happen before validation webhooks are called?

Reply:
@maxsmythe If we use constraint GKNN as the reference object UID, then we will hit the 10 similar events in 10mins `AggregateMaxEvents` limit if there are more than 10 failed admissions for the same constraint. The resulting behavior is that the events with the annotations we generated will get aggregated/clobbered by a new event object that looks something like this:

```sh
39s         Warning   FailedAdmission   namespace/test      (combined from similar events):  Admission webhook "validation.gatekeeper.sh" denied request, Resource Namespace: , Constraint: ns-must-have-gk, Message: you must provide labels: {"gatekeeper"}
```

It will also hit the 25 `SpamBurst` limit pretty fast filtering out new events generated by failed admission requests.
Reference here:
https://github.com/kubernetes/kubernetes/blob/82baa26905c94398a0d19e1b1ecf54eb8acb6029/staging/src/k8s.io/client-go/tools/record/events_cache.go#L414-L429

This can be feature gated with the `emit-deny-events` flag as an opt-in option.

Reply:
@ritazh @maxsmythe Regd. spamming point mentioned above -> As per my understanding authorization check and object schema validation check happens before validation webhooks are called. Is there a scenario where events can be spammed?

Reply:
Depends on how you define authorization.

If someone is using gatekeeper as a way of checking the requesting user (not recommended, but definitely done), then the validation check is part of the authorization process.

Another thing to note is that users could use this to create events without explicitly creating objects, bypassing any quota-style controls that may be in place on the cluster.

To be blunt, completely bypassing all throttling is a bad idea in my opinion, as it offloads all scaling questions to the API server itself, making it more likely that Gatekeeper issues and/or users mishandling Gatekeeper can bring down a cluster.

I am okay with experimenting with this by flag gating it as maybe I'm wrong or maybe its a good way to start a conversation about improving the flexibility of events, but bypassing all throttling entirely is a big red flag and I cannot endorse enabling it by default.

Reply:
At least not without in-the-wild testing, which would necessitate flag-gating anyway.

Reply:
+1 on feature gating this so we can test this feature under heavy load/large cluster. This PR current disables this feature by default.